File: sources/rust/05_alloc_churn.rs
// task 05 alloc_churn — expected output: 1274991808
// build: rustc -O -o prog 05_alloc_churn.rs    run: ./prog

fn main() {
    let mut total: u64 = 0;
    let mut slots: [Option<Box<[u8; 64]>>; 256] = std::array::from_fn(|_| None);

    for i in 0..10_000_000u64 {
        let mut buf = Box::new([0u8; 64]);
        buf[0] = (i % 256) as u8;
        total += u64::from(buf[0]);
        // storing into the slot drops the buffer it replaces, which frees it
        slots[(i % 256) as usize] = Some(buf);
    }

    println!("{}", total);
}

File: sources/objectivec/05_alloc_churn.m
// task 05 alloc_churn -- expected output: 1274991808
// build: clang -fobjc-runtime=gnustep-2.2 -O2 -o prog 05_alloc_churn.m -lobjc -lgnustep-base    run: ./prog
// Objective-C has no big integers in its standard library; task 10 hand-rolls base-1e9 limbs.
#import <Foundation/Foundation.h>
int main(void) { @autoreleasepool {
    long long total = 0;
    unsigned char *slots[256];
    for (int i = 0; i < 256; i++) slots[i] = NULL;
    for (long long i = 0; i < 10000000LL; i++) {
        unsigned char *buf = malloc(64);
        buf[0] = (unsigned char)(i % 256);
        total = total + buf[0];
        int slot = (int)(i % 256);
        if (slots[slot]) free(slots[slot]);
        slots[slot] = buf;
    }
    for (int i = 0; i < 256; i++) free(slots[i]);
    printf("%lld\n", total);
} return 0; }
